libkbfs/impatient_debug_dumper.go

```diff
@@ -0,0 +1,162 @@
+	}
+	return c.front.tracker
+}
+
+func (c *ctxTimeTrackerList) popFront() (t *ctxTimeTracker) {
+	if c.front == nil {
+		return nil
+	}
+	t = c.front.tracker
+	c.front = c.front.next
+	if c.front == nil {
+		// last one!
+		c.tail = nil
+	}
+	return t
+}
+
+func (c *ctxTimeTrackerList) append(t *ctxTimeTracker) {
+	newNode := &ctxTimeTrackerListNode{
+		tracker: t,
+	}
+	if c.tail != nil {
+		c.tail.next = newNode
+	}
+	c.tail = newNode
+	if c.front == nil {
+		c.front = newNode
+	}
+}
+
+type ImpatientDebugDumper struct {
+	config Config
+	logger logger.Logger
+	dumpIn time.Duration
+
+	ticker  *time.Ticker
+	limiter *rate.Limiter
+
+	lock          sync.Mutex
+	chronological *ctxTimeTrackerList
+}
+
+const impatientDebugDumperCheckInterval = time.Second
+const impatientDebugDumperDumpMinInterval = time.Second // 1 dump per sec max
+
+func NewImpatientDebugDumper(config Config, dumpIn time.Duration) *ImpatientDebugDumper {
+	d := &ImpatientDebugDumper{
+		config: config,
+		logger: config.MakeLogger("IGD"),
+		dumpIn: dumpIn,
+		ticker: time.NewTicker(impatientDebugDumperCheckInterval),
+		limiter: rate.NewLimiter(
+			rate.Every(impatientDebugDumperDumpMinInterval), 1),
+		chronological: &ctxTimeTrackerList{},
+	}
+	go d.dumpLoop()
+	return d
+}
+
+func (d *ImpatientDebugDumper) dump(tracker *ctxTimeTracker) {
+	if !d.limiter.Allow() {
+		// Use a limiter to avoid dumping too much into log accidently.
+		return
+	}
+	buf := &bytes.Buffer{}
+	base64er := base64.NewEncoder(base64.StdEncoding, buf)
+	gzipper := gzip.NewWriter(base64er)
+	for _, p := range pprof.Profiles() {
+		fmt.Fprintf(gzipper,
+			"\n======== START Profile: %s ========\n\n", p.Name())
+		_ = p.WriteTo(gzipper, 2)
+		fmt.Fprintf(gzipper,
+			"\n======== END   Profile: %s ========\n\n", p.Name())
+	}
+	gzipper.Close()
+	base64er.Close()
+	d.logger.CDebugf(tracker.ctx,
+		"operation expired. dump>gzip>base64: %q", buf.String())
+}
+
+func (d *ImpatientDebugDumper) dumpTick() {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	for {
+		t := d.chronological.peekFront()
+		if t == nil {
+			return
+		}
+		if atomic.LoadInt32(&t.done) == 1 {
+			// This operation is done, so just move on.
+			d.chronological.popFront()
+			continue
+		}
+		if d.config.Clock().Now().After(t.expiresAt) {
+			// This operation isn't done, and it has expired. So dump debug
+			// information and move on.
+			d.dump(t)
+			d.chronological.popFront()
+			continue
+		}
+		// This operation isn't done yet, but it also hasn't expired. So
+		// just return and wait for next tick and check again.
+		return
+	}
+}
+
+func (d *ImpatientDebugDumper) dumpLoop() {
+	for range d.ticker.C {
+		d.dumpTick()
+	}
+}
+
+func (d *ImpatientDebugDumper) Begin(ctx context.Context) (done func()) {
+	tracker := &ctxTimeTracker{
+		ctx:       ctx,
+		expiresAt: d.config.Clock().Now().Add(d.dumpIn),
+	}
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	d.chronological.append(tracker)
+	return func() {
+		atomic.StoreInt32(&tracker.done, 1)
+	}
```

Comment:
This should return a `*ctxTimeTracker`, whose `Finish` method can be called afterwards instead of a `func()`.

Reply:
Hmmm I found returning a function more clear in its mission. Is there a reason for returning the pointer to the whole thing?

Reply:
It hides concerns about underlying implementation. Forcing another component to be aware of `atomic.StoreInt32` is abstraction-breaking.

Reply:
Why is it forcing another component to use `atomic`? The caller just gets a function to call in a `defer` statement right?

Reply:
Oh you are saying the `atomic` shouldn't be used in this function? I guess I can have a `Finish` and just return `tracker.Finish` here

Reply:
That would work too, although I don't really see why returning the tracker is so bad.

Reply:
You could also make a `tracker.Close()`, and then return an `io.Closer`.

Reply:
@jzila: I don't follow what's wrong with the closure as is.  It doesn't leak anything to the calling components, it doesn't even take any arguments...

Reply:
1. `ImpatientDebugDumper` is having to be aware of the underlying implementation of `ctxTimeTracker`.
2. IMO `return t` then `defer t.Close()` or `return t.Close` then `defer closer()` is much clearer than `return func() { atomic.StoreInt32(&tracker.done, 1) }`. Likewise, in `dumpTick`, `t.IsDone()` is clearer than `atomic.LoadInt32(&t.done) == 1`.

Reply:
Oh, I guess I consider `ctxTimeTracker` to be a sub-type of `ImpatientDebugDumper`.  I thought you were talking about the users of `ImpatientDebugDumper`.  I have no opinion on the internal cleanliness, both are fine with me.